MOL-34444: upgrade to ESLint 9

Txabox opened the pull request:
Migrate eslint-config-ts-mailonline from ESLint 8 to ESLint 9 with the new flat config format.

Major changes:
- Upgrade node version in .nvmrc from 20.6.1 to lts/jod (v22)
- Convert package to ES modules (`"type": "module"`)
- Replace `.eslintrc.*` configuration with flat config (`eslint.config.js`)
- Add named exports (`baseConfig`, `reactConfig`) replacing the previous `extends` approach
- Use `projectService` for automatic tsconfig discovery (replaces manual `project` option)
- Update all dependencies to ESLint 9 compatible versions:
  - eslint: ^8.56.0 → ^9.38.0
  - @typescript-eslint/*: ^6.19.1 → ^8.46.2
  - eslint-plugin-unicorn: ^50.0.1 → ^62.0.0
  - eslint-plugin-react: ^7.33.2 → ^7.37.5
  - eslint-config-prettier: ^9.1.0 → ^10.1.8
  - prettier: ^2.4.1 → ^3.6.2
  - husky: ^8.0.3 → ^9.1.7
  - pretty-quick: ^3.1.1 → ^4.2.2
  - typescript: 5.1.6 → ^5.9.3

BREAKING CHANGE: Requires ESLint >= 9.0.0, Node.js >= 18.18.0, TypeScript >= 5.0.0, and Prettier >= 3.0.0
BREAKING CHANGE: Configuration format changed from `.eslintrc.*` to `eslint.config.js` (flat config)
BREAKING CHANGE: Usage changed from `extends: ["ts-mailonline"]` to `import {baseConfig} from 'eslint-config-ts-mailonline'`
BREAKING CHANGE: `.eslintignore` is no longer supported; ignores must be defined in the config file
BREAKING CHANGE: `tsconfig.eslint.json` is no longer required; the config now uses `projectService` for automatic discovery
BREAKING CHANGE: Rule `@typescript-eslint/ban-types` replaced with `@typescript-eslint/no-empty-object-type`
BREAKING CHANGE: Rule `@typescript-eslint/no-empty-interface` replaced with `@typescript-eslint/no-empty-object-type`

Issues: [MOL-34444](https://anm508a.atlassian.net/browse/MOL-34444)

[MOL-34444]: https://anm508a.atlassian.net/browse/MOL-34444?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Txabox committed:
[chore]: upgrade to ESLint 9

Migrate eslint-config-ts-mailonline from ESLint 8 to ESLint 9 with the new flat config format.

Major changes:
- Upgrade node version in .nvmrc from 20.6.1 to lts/jod (v22)
- Convert package to ES modules (`"type": "module"`)
- Replace `.eslintrc.*` configuration with flat config (`eslint.config.js`)
- Add named exports (`baseConfig`, `reactConfig`) replacing the previous `extends` approach
- Use `projectService` for automatic tsconfig discovery (replaces manual `project` option)
- Update all dependencies to ESLint 9 compatible versions:
  - eslint: ^8.56.0 → ^9.38.0
  - @typescript-eslint/*: ^6.19.1 → ^8.46.2
  - eslint-plugin-unicorn: ^50.0.1 → ^62.0.0
  - eslint-plugin-react: ^7.33.2 → ^7.37.5
  - eslint-config-prettier: ^9.1.0 → ^10.1.8
  - prettier: ^2.4.1 → ^3.6.2
  - husky: ^8.0.3 → ^9.1.7
  - pretty-quick: ^3.1.1 → ^4.2.2
  - typescript: 5.1.6 → ^5.9.3

BREAKING CHANGE: Requires ESLint >= 9.0.0, Node.js >= 18.18.0, TypeScript >= 5.0.0, and Prettier >= 3.0.0
BREAKING CHANGE: Configuration format changed from `.eslintrc.*` to `eslint.config.js` (flat config)
BREAKING CHANGE: Usage changed from `extends: ["ts-mailonline"]` to `import {baseConfig} from 'eslint-config-ts-mailonline'`
BREAKING CHANGE: `.eslintignore` is no longer supported; ignores must be defined in the config file
BREAKING CHANGE: `tsconfig.eslint.json` is no longer required; the config now uses `projectService` for automatic discovery
BREAKING CHANGE: Rule `@typescript-eslint/ban-types` replaced with `@typescript-eslint/no-empty-object-type`
BREAKING CHANGE: Rule `@typescript-eslint/no-empty-interface` replaced with `@typescript-eslint/no-empty-object-type`

Issues: MOL-34444

Txabox committed:
fix: update GitHub Actions to Node 22 and latest action versions

Txabox committed:
fix: remove deprecated husky shebang lines

Txabox committed:
refactor: move config files to src/ folder and add react-hooks plugin

- Move all configuration files (baseConfig.js, reactConfig.js, prettierConfig.js, index.js) to src/ folder
- Add eslint-plugin-react-hooks@^7.0.1 as a dependency
- Include react-hooks rules in reactConfig by default using flat config preset
- Update package.json exports to point to new src/ paths
- Update check-prettier-conflicts scripts to use new file paths
- Update eslint.config.js and prettier.config.js to import from src/
- Update tsconfig.json to include src/**/*.js
- Add proper JSDoc type annotation to prettierConfig.js
- Update README.md and migration docs with new import paths and react-hooks info

BREAKING CHANGE: Import path for prettier config changed from 'eslint-config-ts-mailonline/prettier' to 'eslint-config-ts-mailonline/prettierConfig'
BREAKING CHANGE: eslint-plugin-react-hooks is now included in reactConfig by default - remove duplicate plugin configuration to avoid errors

Issues: MOL-34444

Txabox committed:
docs: update link text for ESLint migration guide in migration-v5.md

Issues: MOL-34444

Txabox committed:
docs: swap links for ESLint migration and configuration guides in migration-v5.md

Issues: MOL-34444

carpasse commented on docs/migration-v5.md:
```diff
@@ -0,0 +1,269 @@
+
+ESLint 9 and the updated plugins may catch issues that weren't detected before. Review and fix any new errors before proceeding.
+
+### 6. Update `prettier.config.js`
```
Suggestion, most of hour projects are commonjs. I would change the extension of this file to `.mjs` to avoid problems

Txabox replied:
same as above

Txabox committed:
chore: add TypeScript declarations and JSDoc documentation

- Add TypeScript declaration generation (.d.ts) for better IntelliSense
- Add comprehensive JSDoc documentation to baseConfig, reactConfig, and prettierConfig
- Update package.json exports with proper types field
- Restore *.config.js to ignores (same behavior as v4). It was bringing problems with tsc builds
- Update migration guide to reflect config files are ignored

Issues: MOL-34444

Txabox committed:
chore: update configuration files to use .mjs extension and add new config files

Issues: MOL-34444

Txabox committed:
chore: add prebuild script to clean dist directory and update rimraf dependency

Issues: MOL-34444

Txabox committed:
chore: update migration guide and examples to use .mjs extension for config files

Issues: MOL-34444

Txabox committed:
chore: update ignore patterns to support .mjs configuration files

Issues: MOL-34444

Txabox committed:
docs: replace `prettier.config.js` with `prettier.config.mjs` in migration guide

Issues: MOL-34444

Txabox committed:
chore: update migration guide and base configuration to require tsconfig.eslint.json

BREAKING CHANGE: now tsconfig.eslint.json is mandatory in parent packages

Issues: MOL-34444

Txabox committed:
docs: update migration guide and README to require tsconfig.eslint.json for type-aware linting

Issues: MOL-34444

Txabox committed:
chore: update include patterns in README and migration guide for new file extensions

Issues: MOL-34444

Txabox committed:
chore: update tsconfig.eslint.json to include additional file extensions for linting

Issues: MOL-34444

Txabox committed:
chore: update ignore patterns to include node_modules in baseConfig and migration guide

Issues: MOL-34444

Txabox committed:
docs: add solution for "No default export found in imported module" errors in migration guide

Issues: MOL-34444

Txabox committed:
fix: downgrade eslint-plugin-unicorn to 59.x to fix parsing error

eslint-plugin-unicorn v60.0.0+ introduced ES2022 string literal exports
syntax (`export {default as 'rule-name'}`) in rules/index.js which causes
parsing errors when eslint-plugin-import's resolver attempts to analyze
files in node_modules using @typescript-eslint/parser.

Error message:
  Error while parsing .../node_modules/eslint-plugin-unicorn/rules/index.js
  Line 3, column 19: Identifier expected.
  `parseForESLint` from parser `context.languageOptions.parser` is invalid

This syntax is valid ES2022 but @typescript-eslint/parser does not support
it when used by eslint-import-resolver-typescript to resolve module imports.

Downgrading to v59.x resolves the issue as it uses standard ESM imports
without string literal export names.

Issues: MOL-34444

Txabox committed:
refactor: update import path for baseConfig in eslint.config.mjs

Issues: MOL-34444

Txabox committed:
chore: remove typescript from peerDependencies

TypeScript is already declared as a peerDependency in @typescript-eslint/parser
(>=4.8.4 <6.0.0), making it redundant to declare it again in this package.

Removing it avoids potential version mismatches in the future, as
@typescript-eslint/parser maintains its own TypeScript version requirements
that may change independently. By delegating this to the parser, consumers
will get the correct version warnings directly from the source of truth.

Issues: MOL-34444

Txabox committed:
chore: add typescript to update dependencies in migration guide

Issues: MOL-34444

Txabox committed:
docs: update README and migration guide to include bundled ESLint plugins and removal instructions

Issues: MOL-34444

Txabox committed:
docs: update migration guide to address deprecation warnings and husky installation changes

Issues: MOL-34444

Txabox committed:
docs: update migration guide to reorder steps and clarify dependency updates

Issues: MOL-34444

Txabox committed:
chore: add typescript to peerDependencies

Issues: MOL-34444

Txabox committed:
chore: update tsconfig and migration guide to exclude coverage directory

Issues: MOL-34444

Txabox committed:
style: enable bracketSpacing for better readability in prettier

Remove explicit `bracketSpacing: false` to use Prettier's default (true).
Objects will now have spaces inside braces: `{ foo: bar }` instead of `{foo: bar}`.

This change applies incrementally via pretty-quick on staged files. So no big-bang changes in PRs, but incremental.

This change improves readability

Issues: MOL-34444

ap2405 reviewed (approved):
Well done Txabi! 🚀

Txabox committed:
docs: update migration guide to include VSCode configuration for ESLint flat config

Issues: MOL-34444

Txabox committed:
chore: update tsconfig to include "noEmit" option for better TypeScript configuration

Issues: MOL-34444

Txabox committed:
refactor: use minimal react-hooks rules instead of recommended preset

Replace eslint-plugin-react-hooks recommended configuration with explicit
rules-of-hooks and exhaustive-deps only.

The recommended preset includes additional rules designed to prepare
codebases for React Compiler adoption. These stricter rules would generate
numerous breaking changes in our large legacy codebase, requiring complex
fixes and regression testing.

Starting with the two essential hooks rules provides:
- Gradual adoption path for projects upgrading from no hooks linting
- Reduced friction when integrating this package into existing projects
- Flexibility to incrementally add rules as non-breaking improvements

This approach avoids potential future breaking changes if the recommended
preset proves too strict for legacy codebases to adopt.

Issues: MOL-34444

Txabox committed:
docs: update examples to use defineConfig and globalIgnores API

Update README.md and migration-v5.md documentation to use the new
ESLint 9 standard APIs:

- Replace spread operator pattern with defineConfig() wrapper
- Use globalIgnores() for file exclusions instead of inline ignores
- Add links to official ESLint documentation

Technical reasons for this change:

1. defineConfig provides better type safety and IDE autocompletion
2. defineConfig validates configuration at definition time, catching
   errors earlier
3. globalIgnores ensures ignore patterns are truly global and properly
   merged with baseConfig ignores, avoiding confusion about scope
4. This aligns with ESLint's official recommended patterns for flat
   config, making it easier for developers to find help in official docs

The documentation now clearly shows both patterns:
- Single config object for simple cases
- Array of config objects when combining globalIgnores with extends

Reference: https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file

Issues: MOL-34444

Txabox committed:
docs: update README to include pre-push hook and clarify husky scripts

Issues: MOL-34444
